### PR TITLE
Incremental refactor of runtime_env for consistency

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Unify @iycheng and @architkulkarni 's PRs to use common code for parsing `runtime_env`. This is an incremental step towards the full runtime_env spec in https://docs.google.com/document/d/1gLGl_eNUsf0O2se4rKg7fbf9IhHtst1w8ydGPNCXnQI/edit#

Followups:
- Replace working_dir / working_dir_uri with `files`.
- Unify env vars vs job_config for specifying runtime config

### CODE REVIEW

File: python/ray/_private/runtime_env.py [hunk 1]
```diff
@@ -24,6 +26,41 @@
...
+class RuntimeEnvDict:
+    """Parses and validates runtime env dictionary from the user."""
+
+    def __init__(self, runtime_env_json: dict):
+        if "conda" in runtime_env_json:
+            self.conda = runtime_env_json["conda"]
+        else:
+            self.conda = None
+        if "files" in runtime_env_json:
+            self.files = runtime_env_json["files"]
```
Comment (edoakes) [inline python/ray/_private/runtime_env.py:38]: as discussed, change to `working_dir` and `py_modules`

File: python/ray/_private/runtime_env.py [hunk 2]
```diff
@@ -24,6 +26,41 @@
...
+        else:
+            self.files = None
+        # TODO(ekl) we should have better schema validation here.
```
Comment (edoakes) [inline python/ray/_private/runtime_env.py:41]: we can use the same as autoscaler for this. i'll file an issue

Comment (edoakes) [reply python/ray/_private/runtime_env.py:41]: https://github.com/ray-project/ray/issues/14635

File: python/ray/_private/runtime_env.py [hunk 3]
```diff
@@ -24,6 +26,41 @@
...
+
+    def to_worker_env_vars(self, override_environment_variables: dict) -> dict:
+        """Given existing worker env vars, return an updated dict.
+
+        This sets any necessary env vars to setup the runtime env.
+        TODO(ekl): env vars is probably not the right long term impl.
```
Comment (edoakes) [inline python/ray/_private/runtime_env.py:47]: yeah let's also just put `env_vars` into runtime_env, maybe in separate PR? up to you

**Review by edoakes — COMMENTED** [review]
Looks good, just some small updates from our discussion earlier today

File: python/ray/_private/runtime_env.py (hunk 3 above)
Comment (ericl) [reply python/ray/_private/runtime_env.py:47]: Added as a TODO for now

File: python/ray/_private/runtime_env.py [hunk 4]
```diff
@@ -24,6 +26,42 @@
...
+
+    def __init__(self, runtime_env_json: dict):
+        if "conda" in runtime_env_json:
+            self.conda = runtime_env_json["conda"]
+        else:
+            self.conda = None
+        if "working_dir" in runtime_env_json:
+            self.working_dir = runtime_env_json["working_dir"]
+        else:
+            self.working_dir = None
```
Comment (iycheng) [inline python/ray/_private/runtime_env.py:40]: nit: should this be clearer? `self.working_dir = runtime_env_json.get("working_dir")`

**Review by fishbone — APPROVED** [review]
LGTM.

File: python/ray/_private/runtime_env.py (hunk 4 above)
Comment (ericl) [reply python/ray/_private/runtime_env.py:40]: Yeah agreed. We'll probably have to rewrite this entire thing though once we add the schema validation.

File: python/ray/_private/runtime_env.py [hunk 5]
```diff
@@ -24,6 +26,71 @@
...
+        # TODO(ekl) we should have better schema validation here.
+        # TODO(ekl) support env_vars, docker, py_modules
```
Comment (edoakes) [inline python/ray/_private/runtime_env.py:71]: FYI https://github.com/ray-project/ray/issues/14650

File: python/ray/_private/runtime_env.py [hunk 6]
```diff
@@ -24,6 +26,71 @@
...
+
+    def to_worker_env_vars(self, override_environment_variables: dict) -> dict:
+        """Given existing worker env vars, return an updated dict.
+
+        This sets any necessary env vars to setup the runtime env.
+        TODO(ekl): env vars is probably not the right long term impl.
```
Comment (edoakes) [inline python/ray/_private/runtime_env.py:77]: yeah we can easily change this to `conda activate` after @architkulkarni adds the "shim process"

**Review by edoakes — APPROVED** [review]

File: python/ray/_private/runtime_env.py [hunk 7]
```diff
@@ -24,6 +26,71 @@
...
+        if override_environment_variables is None:
+            override_environment_variables = {}
+        if self.conda:
+            conda_env_dir = get_conda_env_dir(self.conda)
+            if override_environment_variables is None:
+                override_environment_variables = {}
+            override_environment_variables.update(PYTHONHOME=conda_env_dir)
+        if self.working_dir:
+            if override_environment_variables is None:
+                override_environment_variables = {}
```
Comment (architkulkarni) [inline python/ray/_private/runtime_env.py:88]: Are the inner `if override_environment_variables is None` checks redundant?  But maybe it's safer this way.

File: python/ray/worker.py [hunk 8]
```diff
@@ -1846,7 +1848,7 @@ def method(self):
             infinite retries.
         runtime_env (Dict[str, Any]): Specifies the runtime environment for
             this actor or task and its children.  Currently supports the
-            key "conda_env", whose value should be a string which is the
+            key "conda", whose value should be a string which is the
```
Comment (architkulkarni) [inline python/ray/worker.py:1851]: Should we replace this with  `A runtime environment dictionary (see``runtime_env.py`` for detailed documentation)` as above?

Comment (architkulkarni) [inline python/ray/worker.py:1851]: Should we replace this with  `A runtime environment dictionary (see``runtime_env.py`` for detailed documentation)` as above?

**Review by architkulkarni — APPROVED** [review]
Looks great, thanks for this! I recorded some nits but I can address them in my next PR

File: python/ray/_private/runtime_env.py (hunk 6 above)
Comment (ericl) [reply python/ray/_private/runtime_env.py:77]: I see, the shim process should probably be using @iycheng 's new general purpose io workers to schedule misc work in python processes.

**Comment by ericl** [comment]
//python/ray/tests:test_runtime_env                                      FAILED in 3 out of 3 in 588.4s

**Pull request merged**